content/200-concepts/100-components/02-prisma-client/260-client-extensions/index.mdx

```diff
@@ -0,0 +1,170 @@
+
+// Declare extension B
+const extendsB = prisma.$extends(extensionB)
+
+// Declare an extension that is a combination of A and B
+const extendsAB = prisma.$extends(extensionA).$extends(extensionB)
+```
+
+When you combine two or more extensions, then the last extension declared takes precedence in any conflict. In the example above, suppose there is a method called `getName` in `extendsA` and a method called `getName` in `extendsB`. When you call `extendsAB.getName`, then client uses `getName` as defined in `extendsB`.
+
+### Extensions versus middlewares
+
+Extensions differ from [middlewares](/concepts/components/prisma-client/middleware#performance-and-appropriate-use-cases) in the following ways:
+
+- If you use two or more middlewares at the same time, then they all apply globally to the same client
+- Extensions are isolated: the functionality in each one is separate, unless you deliberately combine them
+- All extended clients share the same connection pool, but their functionality is separate
+- In the RLS example given above, with extensions each user can be in an entirely separate client. With middlewares, all users are active in the same client.
+
+## Share Prisma Client extensions
+
+You can share Prisma Client extensions with other users, either as packages or as code snippets. To create an extension that is more likely to work for other users, use the `$all` prefix, for example [`$allModels`](/concepts/components/prisma-client/client-extensions/client-extensions-model#add-a-custom-method-to-all-models-in-your-schema). This extension applies generally, not just to your specific schema.
+
+For example, a generic extension might take the following form:
+
+```ts
+const xprisma = new prisma.$extends({
+  model: {
+    $allModels: {
+      getClass<T extends object>(this: T): new () => T {
+        return class {
+          /** Generic Implementation */
+        } as any
+      },
+    },
+  },
+})
+
+class UserService extends xprisma.user.getClass() {
+  method() {
+    const user = this.findFirst({}) // fully type-safe
+  }
+}
+```
+
+### Package an extension
+
+To package an extension, use `Prisma.createExtension`. This is very similar to `prisma.$extends`, but it declares an extension.
+For example:
+
+```ts
+
+import { Prisma } from '@prisma/client'
+
+export default Prisma.createExtension({
+  model: {
+    $allModels: {
+      // new method
+      findOrCreate(...) { } // code for the new method goes inside the brackets
+    }
+  }
+})
+```
+
+You could then share the above code on `npm` or as a code snippet.
+
+### Install a packaged extension
+
+In your project, you can install any Prisma Client extension that another user has published to `npm`. To do so, add the following line to your application code:
+
+```ts
+npm install prisma-extension-<name>
+```
+
+For example, if an available extension is called `find-or-create`, you could install it as follows:
+
+```ts
+npm install prisma-extension-find-or-create
```

Comment:
I think there is not really a tight coupling between npm package name and extension name, right? This _can_ be the case, but can also be completely different and random.

Reply:
I'll ask Pierre and Serhii about this